File: fun_matlab/fun_PartExtrap.m
%% Partition extraction algorithm calling Levin-Sidi rule
function [val] = fun_PartExtrap(f,a,q,tol)
%% Initialization
kmax=10; 
eps=1e-6;
old=1e32; 
u=0; w=0; s=0;
A=zeros(kmax+1,1); B=zeros(kmax+1,1);
X=zeros(kmax+2,1);
X(1)=a;
%% Begin extrapolated integration over partitions
for kk=2:kmax+2
   X(kk)=X(kk-1)+q;
   u=fun_TanhSinh(f,X(kk-1),X(kk),eps);
%    if abs(u)==0 a che serve?!
%    end
   %% Execute extrapolation
   s=s+u;
   w=u; %Type of levin transformation
   [val,A,B]=fun_LevinSidi(kk,s,w,X,A,B);
   if kk>3 && abs(val-old) < tol*abs(val)
       break;
   end
   old=val;
end
%check with quadk
% val
% sum=0;
% for ii=2:kk
%     [result(ii-1)]=fun_GaussKronrodBoost(f,X(kk-1),X(kk),eps);
%     sum=sum+result(ii-1);
% end
% sum
if isnan(val) || isinf(val)
 val=fun_GaussKronrodBoost(f,X(1),X(kk),1e-9);
end
% q = integral(f,X(1),X(kk))
%%
% rr=linspace(X(1),X(kk),300);
% y=abs(f(rr));
% figure
% semilogy(rr,abs(f(rr)),'-xr','MarkerSize',10)
% hold on
% for ii=1:kk
%     xline(X(ii))
% end
% drawnow
% 
% 
% rr=linspace(0,X(kk),3000);
% y=abs(f(rr));
% figure
% semilogy(rr,abs(f(rr)),'-xr','MarkerSize',10)
% hold on
% for ii=1:kk
%     xline(X(ii))
% end

end

File: fun_fortran/fun_PartExtrap.f90
!! Partition extraction algorithm calling Levin-Sidi rule
complex*16 function fun_PartExtrap(a,q,tol,component,loc_r1,loc_r,loc_e,loc_k,loc_freq,loc_rho,loc_order,loc_kmax) result(y)
use myfargs
implicit none
!Input
real*8, intent(in) :: a, q, tol
integer, intent(in) ::  component, loc_kmax
real*8, intent(in) ::  loc_order
real*8, intent(in) :: loc_r1(3), loc_r(3)
complex*16, intent(in) :: loc_e(2), loc_k(2)
real*8, intent(in) :: loc_freq, loc_rho
!User
complex*16 AA(loc_kmax+1), BB(loc_kmax+1)
real*8 XX(loc_kmax+2)
integer kk
real*8 eps
complex*16 old, u ,w, s
complex*16, external :: fun_TanhSinh
complex*16, external :: fun_GaussKronrodBoost
! Initialize
eps=1e-6
AA(1:loc_kmax+1)=cmplx(0.0,0.0)
BB(1:loc_kmax+1)=cmplx(0.0,0.0)
XX(1:loc_kmax+2)=0.d0
old=cmplx(1e32,0.0)
u=cmplx(0.0,0.0)
w=cmplx(0.0,0.0)
s=cmplx(0.0,0.0)
XX(1)=a
!
!Begin extrapolated integration over partitions
do kk=2,loc_kmax+2
   XX(kk)=XX(kk-1)+q
   u=fun_TanhSinh(XX(kk-1),XX(kk),eps,component,loc_r1,loc_r,loc_e,loc_k,loc_freq,loc_rho,loc_order)
    !write(6,*) 'XK(kk)', XX(kk)
   !Execute extrapolation
   s=s+u
   w=u !Type of levin transformation
   
   call fun_LevinSidi(kk,s,w,XX,loc_kmax,AA,BB,y)
    !write(6,*) 'y', y
    !write(6,*) 'old', old
    
   
   if (kk>3 .AND. abs(y-old) < tol*abs(y)) then
       exit
   end if
   old=y
enddo


if (isnan(abs(y)) .OR. (abs(y)>1.0d12)) then
 y=fun_GaussKronrodBoost(a,XX(kk-1),1d-9,component,loc_r1,loc_r,loc_e,loc_k,loc_freq,loc_rho,loc_order)
endif

end function fun_PartExtrap